packaging/metadata/_validation.py

```diff
@@ -0,0 +1,159 @@
+from __future__ import annotations
+
+import abc
+import re
+from typing import TYPE_CHECKING, Any, Callable, Dict, Generic, Optional, TypeVar, cast
+
+if TYPE_CHECKING:
+    from ._types import Metadata
+
+
+T = TypeVar("T")
+
+
+class lazy_validator(Generic[T]):  # noqa: N801
+
+    # This hack exists to work around https://github.com/python/mypy/issues/708
+    _creator: Callable[[Any], T] | Callable[[Any], T]
+    _raw_name: str
+    _validators: list[Callable[[Any], None]]
+
+    def __init__(
+        self,
+        creator: Callable[[Any], T],
+        *,
+        raw_name: str | None = None,
+        validators: list[Callable[[Any], None]] | None = None,
+    ) -> None:
+        self._creator = creator
+        if raw_name is not None:
+            self._raw_name = raw_name
+        if validators is not None:
+            self._validators = validators
+        else:
+            self._validators = []
+
+    def __set_name__(self, owner: Metadata, name: str) -> None:
+        if not hasattr(self, "_raw_name"):
+            self._raw_name = name
+
+    def __get__(self, obj: Metadata, owner: type[Metadata]) -> T | None:
+        # TypedDict doesn't support variable key names, and Python 3.7 doesn't
+        # support Literal which would let us let it know that this is validated
+        # already to be safe, so we'll cast here to make things work.
+        raw = cast(Dict[str, Any], obj._raw)
+        validated = cast(Dict[str, Optional[T]], obj._validated)
+
+        if self._raw_name not in validated:
+            value = self._validate(raw.get(self._raw_name))
+            validated[self._raw_name] = value
+            del raw[self._raw_name]
+
+        return validated[self._raw_name]
+
+    def __set__(self, obj: Metadata, value: Any) -> None:
+        raw = cast(Dict[str, Any], obj._raw)
+        validated = cast(Dict[str, Optional[T]], obj._validated)
+
+        validated_value = self._validate(value)
+        validated[self._raw_name] = validated_value
+        raw.pop(self._raw_name, None)
+
+    def __delete__(self, obj: Metadata) -> None:
+        raw = cast(Dict[str, Any], obj._raw)
+        validated = cast(Dict[str, Optional[T]], obj._validated)
+
+        raw.pop(self._raw_name, None)
+        validated.pop(self._raw_name, None)
+
+    def _validate(self, data: Any) -> T | None:
+        # Create our value from our raw data
+        value = self._creator(data) if data is not None else None
+
+        # Loop over our validators, and ensure that our value is actually valid
+        for validator in self._validators:
+            validator(value)
+
+        return value
+
+
+def eagerly_validate(obj: Metadata) -> None:
```

Comment:
We may want to have our own `ExceptionGroup` and have this function raise that for Python < 3.11 so people can know about all validation errors upfront.